x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,162 @@
+    // Having an exponent higher than this causes integer overflow
+    private static final int MAX_RETRY_EXPONENT = 29;
+
+    private final Random random = Randomness.get();
+    private final Client client;
+    private volatile int maxFailureRetries;
+
+    public ResultsPersisterService(Client client, ClusterService clusterService, Settings settings) {
+        this.client = client;
+        this.maxFailureRetries = PERSIST_RESULTS_MAX_RETRIES.get(settings);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(PERSIST_RESULTS_MAX_RETRIES, this::setMaxFailureRetries);
+    }
+
+    void setMaxFailureRetries(int value) {
+        this.maxFailureRetries = value;
+    }
+
+    public BulkResponse indexWithRetry(String jobId,
+                                       String indexName,
+                                       ToXContent object,
+                                       ToXContent.Params params,
+                                       WriteRequest.RefreshPolicy refreshPolicy,
+                                       String id,
+                                       Supplier<Boolean> shouldRetry) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        try (XContentBuilder content = object.toXContent(XContentFactory.jsonBuilder(), params)) {
+            bulkRequest.add(new IndexRequest(indexName).id(id).source(content));
+        }
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry);
+    }
+
+    public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest, String jobId, Supplier<Boolean> shouldRetry) {
+        int currentMin = MIN_RETRY_SLEEP_MILLIS;
+        int currentMax = MIN_RETRY_SLEEP_MILLIS;
+        int currentAttempt = 0;
+        while(currentAttempt <= maxFailureRetries) {
+            BulkResponse bulkResponse = bulkIndex(bulkRequest);
+            if (bulkResponse.hasFailures() == false) {
+                return bulkResponse;
+            }
+            if (shouldRetry.get() == false) {
+                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+            }
+            if (currentAttempt > maxFailureRetries) {
+                LOGGER.warn("[{}] failed to index after [{}] attempts.", jobId, currentAttempt);
+                throw new ElasticsearchException("[{}] failed to index all results after [{}] attempts. {}",
+                    jobId,
+                    currentAttempt,
+                    bulkResponse.buildFailureMessage());
+            }
+            currentAttempt++;
+            // Since we exponentially increase, we don't want force randomness to have an excessively long sleep
+            if (currentMax < MAX_RETRY_SLEEP_MILLIS) {
+                currentMin = currentMax;
+            }
+            double backOff = ((1 << Math.min(currentAttempt, MAX_RETRY_EXPONENT)) - 1) / 2.0;
```

Comment:
Why do you divide by `2.0`?

Reply:
https://en.wikipedia.org/wiki/Exponential_backoff